### PR TITLE
[Frontend] Add LLM.reward specific to reward models

### PR DESCRIPTION
**Pull request by noooop**
## Essential Elements of an Effective PR Description Checklist
- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.

## Purpose

After #21470, LLM.encode pooling_task defaults to using "encode", meaning that reward models actually use LLM.encode. However, most libraries' encode corresponds to an embedding model, which can lead to potential errors.

- Add LLM.reward specified to reward models
- The pooling_task for LLM.encode defaults to None, and pooling_task needs to be specified

### CODE REVIEW

**Comment by noooop** [comment]
@DarkLight1337

Quick review

File: vllm/entrypoints/llm.py [hunk 1]
```diff
@@ -1092,6 +1092,15 @@ def encode(
             considered legacy and may be deprecated in the future. You should
             instead pass them via the `inputs` parameter.
         """
+        if pooling_task is None:
+            raise ValueError(
```
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py]: We should have a soft deprecation period to avoid breaking users' code

Comment (noooop) [reply vllm/entrypoints/llm.py]: how about set the default pooling_task of LLM.encode as embed, add this soft deprecation warning, and migrate all reward models to LLM.reward

Comment (DarkLight1337) [reply vllm/entrypoints/llm.py]: The default should be `reward` since only reward models are documented as using this API

Comment (noooop) [reply vllm/entrypoints/llm.py]: However, most libraries' encode corresponds to an embedding model, which can lead to potential errors.

Comment (noooop) [reply vllm/entrypoints/llm.py]: You can imagine how I discovered this issues

File: docs/models/pooling_models.md [hunk 2]
```diff
@@ -154,6 +139,43 @@ print(f"Score: {score}")
 
 A code example can be found here: <gh-file:examples/offline_inference/basic/score.py>
 
+### `LLM.reward`
+
+The [reward][vllm.LLM.reward] method is available to all pooling models in vLLM.
```
Comment (DarkLight1337) [inline docs/models/pooling_models.md]: I think this is somewhat confusing. People will wonder why is it available to all pooling models when it's called `reward`.

Comment (DarkLight1337) [reply docs/models/pooling_models.md]: Please also update the table in `Pooling Tasks`

Comment (noooop) [reply docs/models/pooling_models.md]: I think it would be best to rename encode in vllm.task:PoolingTask to reward, but it requires too many changes.

do you ok with that?

Comment (DarkLight1337) [reply docs/models/pooling_models.md]: I am fine with renaming the task

Comment (DarkLight1337) [reply docs/models/pooling_models.md]: But should we still keep a separate "encode" task?

Comment (noooop) [reply docs/models/pooling_models.md]: I'm also thinking about this issue.

If we don't rename 'encode', the Pooling Tasks in the document will be very confusing. 

I find it hard to explain clearly that the reward model needs to use pooling_task = encode. 

We can first remove pooling_task = encode, and add it back when needed.

Comment (DarkLight1337) [reply docs/models/pooling_models.md]: Hmm, how about we just keep everything as it is, but in the docs we add a warning so users don't confuse `embed` with `encode`?

Comment (noooop) [reply docs/models/pooling_models.md]: Don't touch the encode, otherwise the table above will look strange.

```
| Architecture                                    | `--convert` | Supported pooling tasks       |
|-------------------------------------------------|-------------|-------------------------------|
| `*ForTextEncoding`, `*EmbeddingModel`, `*Model` | `embed`     | `encode`, `embed`             |
| `*For*Classification`, `*ClassificationModel`   | `classify`  | `encode`, `classify`, `score` |
| `*ForRewardModeling`, `*RewardModel`            | `reward`    | `encode`                      |
```

-> 

```
| Architecture                                    | `--convert` | Supported pooling tasks       |
|-------------------------------------------------|-------------|-------------------------------|
| `*ForTextEncoding`, `*EmbeddingModel`, `*Model` | `embed`     | `reward`, `embed`             |
| `*For*Classification`, `*ClassificationModel`   | `classify`  | `reward`, `classify`, `score` |
| `*ForRewardModeling`, `*RewardModel`            | `reward`    | `reward`       
```

Let's figure out how to explain that the reward model needs to use pooling_task = encode.

Comment (noooop) [reply docs/models/pooling_models.md]: Also, I found that

TaskOption = Literal["auto", "generate", "embedding", "embed", "classify",
                     "score", "reward", "transcription", "draft"]

_ResolvedTask = Literal["generate", "transcription", "encode", "embed",
                        "classify", "reward", "draft"]


TaskOption reward will go to _ResolvedTask encode, that's too complicated

But it seems there's no better way

File: docs/models/pooling_models.md [hunk 3]
```diff
@@ -154,6 +139,44 @@ print(f"Score: {score}")
...
+The [encode][vllm.LLM.encode] method is available to all pooling models in vLLM.
+It returns the extracted hidden states directly.
+
+!!! note
+    `LLM.encode` defaults to using `pooling_task = embed`.
+    Please use one of the more specific methods instead of `encode`:
+    - For embeddings, use `LLM.embed(...)`.
+    - For classification logits, use `LLM.classify(...)`.
+    - For reward scores, use `LLM.reward(...)`.
+    - For similarity scores, use `LLM.score(...)`.
```
Comment (noooop) [inline docs/models/pooling_models.md]: @DarkLight1337 

Let's go back to the most important changes

`LLM.encode` defaults to using `pooling_task = embed`

do you ok with that.

Comment (DarkLight1337) [reply docs/models/pooling_models.md]: There are people who are using `LLM.encode` for reward models. so we can't change this immediately. I'm fine about changing this after the deprecation period though.

File: vllm/entrypoints/llm.py [hunk 4]
```diff
@@ -1092,6 +1092,21 @@ def encode(
...
+
+            logger.warning_once(
+                "`LLM.encode` is currently using `pooling_task = %s`.\n"
+                "Please use one of the more specific methods instead "
+                "of `encode`:\n"
+                "  - For embeddings, use `LLM.embed(...)`.\n"
+                "  - For classification logits, use `LLM.classify(...)`.\n"
+                "  - For reward scores, use `LLM.reward(...)`.\n"
+                "  - For similarity scores, use `LLM.score(...)`.", pooling_task)
+
```
Comment (noooop) [inline vllm/entrypoints/llm.py]: how about adding a test if self.llm_engine.model_config.encoder_config is not None, which would make both embed and reward  happy (to the maximum extent possible)

Comment (DarkLight1337) [reply vllm/entrypoints/llm.py]: I think this will just confuse people even more as to what's the meaning of `encode`. Maybe we should just rename `encode` to `pool` to be directly compared to `generate`

Comment (noooop) [reply vllm/entrypoints/llm.py]: > I think this will just confuse people even more as to what's the meaning of `encode`. Maybe we should just rename `encode` to `pool` to be directly compared to `generate`

How about renaming `encode` to `pool` in the next PR, I find it more complex than I thought.

File: vllm/entrypoints/llm.py [hunk 5]
```diff
@@ -1092,6 +1092,22 @@ def encode(
...
+                pooling_task = "encode"
+
+            logger.warning_once(
+                "`LLM.encode` is currently using `pooling_task = %s`.\n"
+                "Please use one of the more specific methods instead "
+                "of `encode`:\n"
+                "  - For embeddings, use `LLM.embed(...)`.\n"
+                "  - For classification logits, use `LLM.classify(...)`.\n"
+                "  - For reward scores, use `LLM.reward(...)`.\n"
+                "  - For similarity scores, use `LLM.score(...)`.",
```
Comment (DarkLight1337) [inline vllm/entrypoints/llm.py]: Now we have a problem that both `reward` and `score` refer to scores

Comment (noooop) [reply vllm/entrypoints/llm.py]: how about 

> - For rewards, use `LLM.reward(...)`.

Comment (DarkLight1337) [reply vllm/entrypoints/llm.py]: Sounds good

File: docs/models/pooling_models.md [hunk 6]
```diff
@@ -154,6 +139,45 @@ print(f"Score: {score}")
 
 A code example can be found here: <gh-file:examples/offline_inference/basic/score.py>
 
+### `LLM.reward`
+
+The [reward][vllm.LLM.reward] method is available to all pooling models in vLLM.
+It returns the extracted hidden states directly, which is useful for reward models.
```
Comment (DarkLight1337) [inline docs/models/pooling_models.md]:
```suggestion
The [reward][vllm.LLM.reward] method is available to all reward models in vLLM.
It returns the extracted hidden states directly.
```

File: docs/models/pooling_models.md [hunk 7]
```diff
@@ -68,7 +68,7 @@ the pooler assigned to each task has the following attributes by default:
 
 | Task       | Pooling Type   | Normalization | Softmax |
 |------------|----------------|---------------|---------|
-| `encode`   | `ALL`          | ❌            | ❌      |
+| `reward`   | `ALL`          | ❌            | ❌      |
```
Comment (DarkLight1337) [inline docs/models/pooling_models.md]: Also fix the table above this

File: docs/models/pooling_models.md [hunk 8]
```diff
@@ -154,6 +139,45 @@ print(f"Score: {score}")
...
+    - For embeddings, use `LLM.embed(...)`.
+    - For classification logits, use `LLM.classify(...)`.
+    - For rewards, use `LLM.reward(...)`.
+    - For similarity scores, use `LLM.score(...)`.  
+
+```python
+from vllm import LLM
+
+llm = LLM(model="intfloat/e5-small", runner="pooling")
+(output,) = llm.encode("Hello, my name is")
```
Comment (DarkLight1337) [inline docs/models/pooling_models.md]:
```suggestion
(output,) = llm.encode("Hello, my name is", task="embed")
```

File: docs/models/pooling_models.md [hunk 9]
```diff
@@ -154,6 +139,45 @@ print(f"Score: {score}")
...
+
+The [encode][vllm.LLM.encode] method is available to all pooling models in vLLM.
+It returns the extracted hidden states directly.
+
+!!! note
+    Please use one of the more specific methods instead of `encode`:
+    - For embeddings, use `LLM.embed(...)`.
+    - For classification logits, use `LLM.classify(...)`.
+    - For rewards, use `LLM.reward(...)`.
+    - For similarity scores, use `LLM.score(...)`.
```
Comment (DarkLight1337) [inline docs/models/pooling_models.md]:
```suggestion
    Please use one of the more specific methods or set the task directly when using `LLM.encode`:
    - For embeddings, use `LLM.embed(...)` or `task="embed".
    - For classification logits, use `LLM.classify(...)` or `task="classify"`.
    - For rewards, use `LLM.reward(...)` or `task="reward"`.
    - For similarity scores, use `LLM.score(...)`.  
```

File: docs/models/pooling_models.md [hunk 10]
```diff
@@ -66,11 +66,11 @@ you can override some of its attributes via the `--override-pooler-config` optio
 If the model has been converted via `--convert` (see above),
 the pooler assigned to each task has the following attributes by default:
 
-| Task       | Pooling Type   | Normalization | Softmax |
-|------------|----------------|---------------|---------|
-| `encode`   | `ALL`          | ❌            | ❌      |
-| `embed`    | `LAST`         | ✅︎            | ❌      |
-| `classify` | `LAST`         | ❌            | ✅︎      |
+| Task       | Pooling Type | Normalization | Softmax |
```
Comment (DarkLight1337) [inline docs/models/pooling_models.md:69]: I mean the table in L48

File: docs/models/pooling_models.md [hunk 11]
```diff
@@ -47,7 +47,7 @@ enabling the corresponding APIs:
 
 | Task       | APIs               |
 |------------|--------------------|
-| `encode`   | `encode`           |
+| `reward`   | `encode`           |
```
Comment (DarkLight1337) [inline docs/models/pooling_models.md]: Actually it's the API that needs to be renamed, not the task

Comment (noooop) [reply docs/models/pooling_models.md]: i see 

I finally understand the difference between task and APIs here.

Comment (noooop) [reply docs/models/pooling_models.md]: How about putting APIs in the first column?

Comment (DarkLight1337) [reply docs/models/pooling_models.md]: This section is talking about the task so task should be in the first column

File: docs/models/pooling_models.md [hunk 12]
```diff
@@ -66,11 +66,11 @@ you can override some of its attributes via the `--override-pooler-config` optio
...
-| Task       | Pooling Type   | Normalization | Softmax |
-|------------|----------------|---------------|---------|
-| `encode`   | `ALL`          | ❌            | ❌      |
-| `embed`    | `LAST`         | ✅︎            | ❌      |
-| `classify` | `LAST`         | ❌            | ✅︎      |
+| Task       | Pooling Type | Normalization | Softmax |
+|------------|--------------|---------------|---------|
+| `reward`   | `ALL`        | ❌             | ❌       |
+| `embed`    | `LAST`       | ✅︎            | ❌       |
+| `classify` | `LAST`       | ❌             | ✅︎      |
```
Comment (DarkLight1337) [inline docs/models/pooling_models.md]:
```suggestion
| `reward`   | `ALL`        | ❌            | ❌     |
| `embed`    | `LAST`       | ✅︎            | ❌      |
| `classify` | `LAST`       | ❌            | ✅︎      |
```

File: docs/models/pooling_models.md [hunk 13]
```diff
@@ -154,6 +139,45 @@ print(f"Score: {score}")
...
+
+A code example can be found here: <gh-file:examples/offline_inference/basic/reward.py>
+
+### `LLM.encode`
+
+The [encode][vllm.LLM.encode] method is available to all pooling models in vLLM.
+It returns the extracted hidden states directly.
+
+!!! note
+    Please use one of the more specific methods or set the task directly when using `LLM.encode`:
```
Comment (DarkLight1337) [inline docs/models/pooling_models.md:165]: The bulleted list here is not formatted properly

**Review by DarkLight1337 — APPROVED** [review]
Looks good now, thanks

**Comment by noooop** [comment]
Thanks for reviewing

**Comment by DarkLight1337** [comment]
Please fix the CI failure

**Comment by DarkLight1337** [comment]
I think we should still keep `encode` in `VllmRunner` just like for `LLM` class

**Comment by noooop** [comment]
> I think we should still keep `encode` in `VllmRunner` just like for `LLM` class

Perhaps we need to figure out whether it requires embed or reward.

**Comment by DarkLight1337** [comment]
This model uses pooler interface for image segmentation. So it's not really embed/reward. The more generic `encode` (which we can rename later)  is a better match.

**Comment by noooop** [comment]
I'll wait for buildkite/ci/pr/language-models-test-extended-pooling to finish and check for any other errors

**Comment by DarkLight1337** [comment]
Let's wait for #18321 to be merged first, then update this PR

**Comment by noooop** [comment]
[buildkite/ci/pr/language-models-test-extended-pooling](https://buildkite.com/vllm/ci/builds/25262#019855cb-f1b2-4cc9-8259-704e6706c33c) 

The failed test can be fixed by #21747

**Comment by xsank** [comment]
@noooop @DarkLight1337 For online server, which api is the best entrypoint for the reward model?

**Comment by noooop** [comment]
> @noooop @DarkLight1337 For online server, which api is the best entrypoint for the reward model?

https://github.com/vllm-project/vllm/blob/main/vllm/entrypoints/openai/serving_pooling.py

/pooling

***

I personally think that /pooling needs to be refactored to better support the reward models

PTAL #20538

**Comment by xsank** [comment]
> > @noooop @DarkLight1337 For online server, which api is the best entrypoint for the reward model?
> 
> https://github.com/vllm-project/vllm/blob/main/vllm/entrypoints/openai/serving_pooling.py
> 
> /pooling
> 
> I personally think that /pooling needs to be refactored to better support the reward models
> 
> PTAL #20538

Now i'm using embedding api and i have no scenario for pooling api yet

**Comment by noooop** [comment]
> > > @noooop @DarkLight1337 For online server, which api is the best entrypoint for the reward model?
> > 
> > 
> > https://github.com/vllm-project/vllm/blob/main/vllm/entrypoints/openai/serving_pooling.py
> > /pooling
> > I personally think that /pooling needs to be refactored to better support the reward models
> > PTAL #20538
> 
> Now i'm using embedding api and i have no scenario for pooling api yet

According to vllm's taxonomy

- the reward model means using all pooling (and step pooling, let's temporarily not discuss it. ), meaning there is an output for each token.

- the embedding model means using [CLS], LAST, MEAN pooling, which means there is only one output for a sequence
